nullaway/src/main/java/com/uber/nullaway/handlers/contract/fieldcontract/EnsuresNonNullIfHandler.java

```diff
@@ -0,0 +1,303 @@
+
+  // The MethodTree and Symbol of the EnsureNonNullIf method under semantic validation
+  @Nullable private MethodTree methodTreeUnderAnalysis;
+  @Nullable private Symbol.MethodSymbol methodSymbolUnderAnalysis;
+
+  public EnsuresNonNullIfHandler() {
+    super("EnsuresNonNullIf");
+  }
+
+  /**
+   * Validates whether all parameters mentioned in the @EnsuresNonNullIf annotation are guaranteed
+   * to be {@code @NonNull} at exit point of this method.
+   */
+  @Override
+  protected boolean validateAnnotationSemantics(
+      MethodTree tree, MethodAnalysisContext methodAnalysisContext) {
+    if (tree.getBody() == null) {
+      return false;
+    }
+
+    // clean up state variables, as we are visiting a new annotated method
+    semanticsHold = false;
+    missingFieldNames = null;
+    methodTreeUnderAnalysis = tree;
+    methodSymbolUnderAnalysis = methodAnalysisContext.methodSymbol();
+    returnTreesInMethodUnderAnalysis.clear();
+
+    VisitorState state = methodAnalysisContext.state();
+    NullAway analysis = methodAnalysisContext.analysis();
+
+    // we visit the tree of the method, just so we can build a map between
+    // return statements and their enclosing methods
+    buildUpReturnToEnclosingMethodMap(state);
+
+    // We force the nullness analysis of the method under validation
+    analysis
+        .getNullnessAnalysis(state)
+        .forceRunOnMethod(new TreePath(state.getPath(), tree), state.context);
+
+    // If listed fields aren't checked for their nullness, return error
+    if (!semanticsHold) {
+      String message;
+      if (missingFieldNames == null) {
+        message = "Method is annotated with @EnsuresNonNullIf but does not return boolean";
+      } else {
+        message =
+            String.format(
+                "Method is annotated with @EnsuresNonNullIf but does not ensure fields %s",
+                missingFieldNames);
+      }
+      state.reportMatch(
+          analysis
+              .getErrorBuilder()
+              .createErrorDescription(
+                  new ErrorMessage(ErrorMessage.MessageTypes.POSTCONDITION_NOT_SATISFIED, message),
+                  tree,
+                  analysis.buildDescription(tree),
+                  state,
+                  null));
+    }
+
+    // Clean up state again
+    semanticsHold = false;
+    missingFieldNames = null;
+    methodTreeUnderAnalysis = null;
+    methodSymbolUnderAnalysis = null;
+    returnTreesInMethodUnderAnalysis.clear();
+
+    return true;
+  }
+
+  private void buildUpReturnToEnclosingMethodMap(VisitorState methodState) {
+    returnTreesInMethodUnderAnalysis.clear();
+    new TreePathScanner<Void, Void>() {
+      @Override
+      public Void visitReturn(ReturnTree node, Void unused) {
+        VisitorState returnState = methodState.withPath(getCurrentPath());
+
+        TreePath enclosingMethod =
+            NullabilityUtil.findEnclosingMethodOrLambdaOrInitializer(returnState.getPath());
```

Comment:
I don't think you need `returnState`?  Just do `NullabilityUtil.findEnclosingMethodOrLambdaOrInitializer(getCurrentPath())`?

Reply:
Yes!